### PR TITLE
Add Device and Utils

### PR DESCRIPTION
**Pull request by iBelieve**
- Add Device and Utils singletons from QuickPaper
- Add initial unit tests for Utils
- Add a Travis CI build config
- Some cleanup to the README and CMake files

A couple of points:
- QuickPaper is licensed under the MPL so it could be used by more people (and on iOS). Is that something you're interested in doing for Fluid?
- The status badges expect a develop branch. That would probably be a good thing to add.

### CODE REVIEW

File: core/device.cpp [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+        return hasTouchScreen() ? 64 : 48;
+    }
+}
+
+void Device::screenChanged()
+{
+    if (m_screen)
+        m_screen->disconnect(this);
+
+    QGuiApplication *app = (QGuiApplication *) QGuiApplication::instance();
```
Comment (plfiorini) [inline core/device.cpp]: No need for cast here (there are other instances)

File: core/device.h [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    QString iconName() const;
+
+    bool isPortrait() const;
+    bool hasTouchScreen() const;
+    bool isMobile() const;
+    bool hoverEnabled() const;
+
+    int gridUnit() const;
+
+signals:
```
Comment (plfiorini) [inline core/device.h]: Q_SIGNALS

File: core/device.h [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    void geometryChanged();
+
+private slots:
```
Comment (plfiorini) [inline core/device.h]: Q_SLOTS

File: core/device.cpp (hunk 1 above)
Comment (iBelieve) [inline core/device.cpp]: The method actually returns a `QCoreApplication *`:

```
/Users/mspencer/Developer/hawaiios/fluid/core/device.cpp:16:22: error: cannot initialize a variable of type 'QGuiApplication *' with an rvalue of type
      'QCoreApplication *'
    QGuiApplication *app = QGuiApplication::instance();
```

File: tests/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,12 @@
+# NOTE: We're not using CTest here as just adds an extra layer around QuickTest
```
Comment (plfiorini) [inline tests/CMakeLists.txt:1]: but CTest adds a make test target that runs all tests, that is helpful for the CI

**Comment by plfiorini** [comment]
What's the difference between LGPL and MPL 2?
How is the combined work of LGPL and MPL code licensed?

File: tests/CMakeLists.txt (hunk 4 above)
Comment (iBelieve) [inline tests/CMakeLists.txt:1]: At least for now all the QML tests are run via one executable, which already displays pass/fail progress, and can be run via `make check`. Using CTest would hide the individual QuickTest results by default and doesn't really provide much advantage now.

**Comment by iBelieve** [comment]
MPL and LGPL would be licensed under the LGPL as a whole. MPL lets developers use the project in closed source projects while requiring that this code remains open source. So if someone makes a change to the project, and publishes a closed source app using the changed library, they have to release the code to their changed version of the library, but not their entire app.

Basically, it allows more people to use the project (good for a general library like this, especially a cross-platform one) while still requiring changes be made public.

File: core/device.h (hunk 2 above)
Comment (iBelieve) [inline core/device.h]: What's the reasoning for using the uppercase macros, when most of the Qt examples (and most code using Qt that I've seen) uses `signals`/`slots`?

**Comment by plfiorini** [comment]
LGPL already allows and mandates that changes to the library be released under the same terms without affecting the application.

File: tests/CMakeLists.txt (hunk 4 above)
Comment (plfiorini) [inline tests/CMakeLists.txt:1]: OK

File: core/device.h (hunk 2 above)
Comment (plfiorini) [inline core/device.h]: avoid conflicts with C headers and the majority of public headers use the upper case notation:

$ grep -rn signals /usr/include/qt5/ | wc -l
50
$ grep -rn Q_SIGNALS /usr/include/qt5/ | wc -l
373

**Comment by iBelieve** [comment]
Right, but there are still some restrictions on the LGPL which make it hard for some projects to use it (particularly those on iOS/macOS, which is important if this is going to be the successor to QML Material). Here's a discussion about licensing and the App Stores: https://forum.syncthing.net/t/discussion-of-license-implications-gpl-lgpl-mpl2-mit-ios-app-stores/2002. Again, not really a big difference between LGPL and MPL on Linux, but thinking about cross-platform, it is a bigger difference.

File: core/device.cpp (hunk 1 above)
Comment (plfiorini) [inline core/device.cpp]: right, use qGuiApp, it's already casted with static_cast

in general avoid C casts and use static_cast in similar cases

File: core/device.h (hunk 2 above)
Comment (iBelieve) [inline core/device.h]: Ok, fixed.

File: core/device.cpp (hunk 1 above)
Comment (iBelieve) [inline core/device.cpp]: Fixed.

**Comment by plfiorini** [comment]
As far as I understand MPL is like LGPL but it doesn't force you to make the library replaceable with a varian like LGPL does. Replacing a library in a app store application is kinda hard.
At this point even Qt cannot be used on iOS app store unless the developers will buy a commercial license.

**Comment by plfiorini** [comment]
As far as I understand MPL (it's a new license for me) seems fine for Fluid (being a cross platform toolkit).
Though I wouldn't use it in any other repository. In the other repositories it should stay GPL/LGPL.

**Comment by plfiorini** [comment]
Merged in newstuff branch because I need to do a 0.8 version based on 0.7 but using Qt 5.7 before officially merge all this new stuff.

**Comment by iBelieve** [comment]
Thanks for the merge. Would it make sense to rename that `newstuff` branch to `develop` and push new features to that, merging into master each time we do a release?

**Comment by plfiorini** [comment]
Current branching model for Hawaii is:
- Development happens in master, should not have build failures
- When a new release is out, master is branched into a versioned branch (for ex. 0.7) where bug fixes can be backported (doesn't have much sense now but after 1.0 is a must)

The develop branch model works like this:
- Development happens in the develop branch (build can be in a broken state from time to time)
- master is the latest known good version and we merge from develop from time to time

Is that right?
Can we still do version named branches?

**Comment by iBelieve** [comment]
@plfiorini Here's a good description of GitFlow: https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow. Typically work happens in the develop branch and then is merged into the master. Releases are usually done as tags on the master branch, so backported fixes would come in the form of hotfixes applied to master and then tagged as a new release. Alternatively, though, we could use version branches if we wanted to backport fixes to something older than the latest stable release. With gitflow, develop should always build successfully, as development takes place in feature branches. However, though develop will always not have build failures, it still could have bugs or other problems not caught by continous integration. Master would have as few bugs as possible (with all bugs fixed as deemed appropriate for a release).

**Comment by plfiorini** [comment]
ok renamed newstuff to develop

**Pull request closed without merging**